vmngclient/api/software_action_api.py

```diff
@@ -0,0 +1,186 @@
+    device_type: DeviceType
+
+
+class SoftwareActionAPI:
+    """
+    API methods for software actions. All methods
+    are exececutable on all device categories.
+    """
+
+    def __init__(self, repository: RepositoryAPI) -> None:
+
+        self.repository = repository
+
+    def activate_software(self, version_to_activate: str) -> str:
+        """
+        Method to set choosen version as current version
+
+        Args:
+            version_to_activate (str): version to be set as current version
+
+        Returns:
+            str: action id
+        """
+        self.repository.complete_device_list(version_to_activate, "available_versions")
+        url = "/dataservice/device/action/changepartition"
+        payload = {
+            "action": "changepartition",
+            "devices": self.repository.devices,
+            "deviceType": "vmanage",
+        }
+        activate = dict(self.repository.session.post_json(url, payload))
+        return activate["id"]
+
+    def upgrade_software(
+        self,
+        software_image: str,
+        install_spec: InstallSpecification,
+        reboot: bool,
+        sync: bool = True,
+    ) -> str:
+        """
+        Method to install new software
+
+        Args:
+            software_image (str): path to software image
+            install_spec (InstallSpecification): specification of devices
+            on which the action is to be performed
+            reboot (bool): reboot device after action end
+            sync (bool, optional): Synchronize settings. Defaults to True.
+
+        Raises:
+            ValueError: Raise error if downgrade in certain cases
+
+        Returns:
+            str: action id
+        """
+        self.install_spec = install_spec
+
+        url = "/dataservice/device/action/install"
+        payload: Dict[str, Any] = {
+            "action": "install",
+            "input": {
+                "vEdgeVPN": 0,
+                "vSmartVPN": 0,
+                "family": self.install_spec.family,
+                "version": self.repository.get_image_version(software_image),
+                "versionType": self.install_spec.version_type,
+                "reboot": reboot,
+                "sync": sync,
+            },
+            "devices": self.repository.devices,
+            "deviceType": self.install_spec.device_type,
+        }
+
+        if self.install_spec.family in (Family.VMANAGE.value, Family.CEDGE.value):
+            incorrect_devices = self._downgrade_check(payload["input"]["version"], self.repository.device_category)
+            if incorrect_devices:
+                raise ValueError(
+                    f"Current version of devices {incorrect_devices} is higher than upgrade version. Action denied!"
+                )
+
+        upgrade = dict(self.repository.session.post_json(url, payload))
+        return upgrade["id"]
+
+    def _downgrade_check(self, version_to_upgrade: str, devices_category: DeviceCategory) -> Union[None, List]:
```

Comment:
I dont really understand what this method does. Also return type is really unclear for me. List of what? Why dont we return empty list? Can you make it more clear

Reply:
Added docstring to method